File: implementations/cpp/null/decode.cpp
#include <cstdint>
#include <cstring>
#include <vector>

#include "../../../harness/cpp/benchmark_harness.hpp"

class NullBench : public BenchmarkImplementation {
 private:
  std::vector<uint8_t> input_data;

 public:
  std::string name() const override { return "null-decode"; }

  void prepare(const Args &args) override {
    input_data = read_binary_file(args.input);
  }

  std::vector<uint8_t> run(const Args &args) override {
    std::vector<uint8_t> output(input_data);

    // Write checksum at end to prevent optimization
    uint32_t checksum = crc32_hash(input_data);
    if (output.size() >= 4) {
      std::memcpy(&output[output.size() - 4], &checksum, 4);
    }

    return output;
  }
};

int main(int argc, char **argv) {
  NullBench bench;
  return run_benchmark(argc, argv, bench);
}

File: implementations/rust/null/src/decode.rs
use anyhow::Result;
use benchmark_harness::{Args, BenchmarkImplementation};
use std::fs;

struct NullBench;

struct BenchContext {
    input_data: Vec<u8>,
}

impl BenchmarkImplementation for NullBench {
    fn name(&self) -> &'static str {
        "null-decode"
    }

    fn prepare(&self, args: &Args) -> Result<Box<dyn std::any::Any>> {
        let input_data = fs::read(&args.input)?;
        Ok(Box::new(BenchContext { input_data }))
    }

    fn run(&self, _args: &Args, context: &mut dyn std::any::Any) -> Result<Vec<u8>> {
        let ctx = context
            .downcast_ref::<BenchContext>()
            .expect("Invalid context");

        let checksum = crc32fast::hash(&ctx.input_data);
        let mut result = vec![0u8; ctx.input_data.len()];
        result.copy_from_slice(&ctx.input_data);

        // Write checksum at end to prevent optimization
        if result.len() >= 4 {
            let len = result.len();
            result[len - 4..].copy_from_slice(&checksum.to_le_bytes());
        }

        Ok(result)
    }
}

fn main() -> Result<()> {
    benchmark_harness::main(NullBench)
}
